sql: implement trigram inverted indexes and index acceleration for % and LIKE

jordanlewis opened the pull request:
Updates #41285

See commits for details. This PR implements inverted indexing on string columns using trigrams and searching trigram indexes using the `=`, `LIKE`, `ILIKE` and `%` (text similarity) operators.

Inverted indexes on text columns are built with the `gin_trgm_ops` or `gist_trgm_ops` opclasses, with `CREATE INDEX ON t USING GIN(col gin_trgm_ops)`, `CREATE INVERTED INDEX ON t(col gin_tgrm_ops)`, or `CREATE INDEX ON t USING GIST(col gist_trgm_ops)`. The `gin` and `gist` opclasses are currently implemented identically.

jordanlewis committed:
sql: implement trigram inverted indexes

This commit implements trigram inverted indexes that are searchable
using the =, LIKE, ILIKE and % (similarity) operators. This is a partial
implementation of the gin_trgm_ops opclass from Postgres. gin_trgm_ops
also supports regex matches in Postgres.

Release note (sql change): string columns now support inverted indexes
using trigrams. These indexes can be searched using the =, LIKE, ILIKE
and % (similarity) predicates.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/79705)
<!-- Reviewable:end -->

jordanlewis commented:
For an example, here's a log from the CLI:

```
demo@127.0.0.1:26257/defaultdb> create table a(a primary key, t) as select 1, 'afooa';
CREATE TABLE AS
demo@127.0.0.1:26257/defaultdb> create inverted index on a(t);
CREATE INDEX
demo@127.0.0.1:26257/defaultdb> explain select * from a where t ilike '%foo%zoo%' OR t ilike '%foo%bar%';
                            info
------------------------------------------------------------
  distribution: local
  vectorized: true

  • filter
  │ filter: (t ILIKE '%foo%zoo%') OR (t ILIKE '%foo%bar%')
  │
  └── • index join
      │ table: a@a_pkey
      │
      └── • inverted filter
          │ inverted column: t_inverted_key
          │ num spans: 3
          │
          └── • scan
                missing stats
                table: a@a_t_idx
                spans: 3 spans
(17 rows)
demo@127.0.0.1:26257/defaultdb> explain select * from a where t ilike '%foo%';
             info
------------------------------
  distribution: local
  vectorized: true

  • filter
  │ filter: t ILIKE '%foo%'
  │
  └── • index join
      │ table: a@a_pkey
      │
      └── • scan
            missing stats
            table: a@a_t_idx
            spans: 1 span
(13 rows)
demo@127.0.0.1:26257/defaultdb> select * from a where t ilike '%foo%';
  a |   t
----+--------
  1 | afooa
(1 row)
demo@127.0.0.1:26257/defaultdb>
```

jordanlewis committed:
sql: permit opclasses for inverted indexes

Previously, it was not permitted to use opclass syntax for index
creation. For example, the `blah_ops` in `CREATE INDEX ON t USING GIN
(col blah_ops))`.

Now, this syntax is legal when the opclass is supported for a given
type.

Release note (sql change): permit usage of jsonb_ops, array_ops,
gin_trgm_ops, and gist_trgm_ops as an opclass in inverted index
creation.

jordanlewis commented:
I added opclass parsing and persistence. I think this RFAL!

awoods187 commented:
This is super cool! What an unexpected win!

jordanlewis commented:
I've split off the builtin functions, operators, session settings, etc into #81418.

jordanlewis committed:
sql: version gate inverted index creation

Make sure that it's not possible to create an inverted index until the
cluster is at a new enough version.

Release note: None

rytaft reviewed (commented):
Awesome! Exciting stuff!

> Reviewed 47 of 47 files at r10, 9 of 9 files at r11, 23 of 23 files at r12, 15 of 15 files at r13, 13 of 13 files at r14, 4 of 4 files at r15, 17 of 17 files at r16, 6 of 6 files at r17, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/79705#-:-N2dQPDVApSGCtO1_5JS:b-2vy6s7)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis)

___
*[`pkg/sql/opt/invertedidx/trigram.go` line 32 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/79705#-N2dC5GB4D9yvJrVqH52:-N2dC5GB4D9yvJrVqH53:b-zav9ao) ([raw file](https://github.com/cockroachdb/cockroach/blob/3c68a70d2624dcb6eca8a5850f3770bfe6a14af6/pkg/sql/opt/invertedidx/trigram.go#L32)):*
> ```Go
> }
> 
> func (t *trigramFilterPlanner) extractInvertedFilterConditionFromLeaf(
> ```

nit: add comment:
```
// extractInvertedFilterConditionFromLeaf is part of the invertedFilterPlanner
// interface.
```
___
*[`pkg/sql/opt/invertedidx/trigram.go` line 81 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/79705#-N2dDKMO0p78kTBjxevL:-N2dDKMO0p78kTBjxevM:bzftbl8) ([raw file](https://github.com/cockroachdb/cockroach/blob/3c68a70d2624dcb6eca8a5850f3770bfe6a14af6/pkg/sql/opt/invertedidx/trigram.go#L81)):*
> ```Go
> }
> 
> var _ invertedFilterPlanner = &trigramFilterPlanner{}
> ```

nit: move this just below the type definition 
___
*[`pkg/sql/opt/invertedidx/trigram.go` line 47 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/79705#-N2dGkE26rp1J5urnDmT:-N2dGkE26rp1J5urnDmU:b-tamwnc) ([raw file](https://github.com/cockroachdb/cockroach/blob/c4e48d1893688bdfcc53592b4927b6e86a0bbd3c/pkg/sql/opt/invertedidx/trigram.go#L47)):*
> ```Go
> 	case *memo.ILikeExpr:
> 		left, right = e.Left, e.Right
> 		// If we're doing a LIKE (or ILIKE) expression, we need to construct an AND out of all
> ```

nit: make comment <= 80 columns wide
___
*[`pkg/sql/opt/invertedidx/trigram.go` line 32 at r17](https://reviewable.io/reviews/cockroachdb/cockroach/79705#-N2dOVgOAmKxCwd3xBz0:-N2dOVgOAmKxCwd3xBz1:b-9nmow4) ([raw file](https://github.com/cockroachdb/cockroach/blob/0d81af2f9e6759aa40f05c5663aa49ebb3e849f4/pkg/sql/opt/invertedidx/trigram.go#L32)):*
> ```Go
> }
> 
> func (t *trigramFilterPlanner) extractInvertedFilterConditionFromLeaf(
> ```

This could use a unit test (similar to `TestTryFilterJsonOrArrayIndex`)
___
*[`pkg/sql/parser/sql.y` line 8710 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/79705#-N2dJpDQEk5pRAbeEiQC:-N2dJpDQEk5pRAbeEiQD:b-r1ywc3) ([raw file](https://github.com/cockroachdb/cockroach/blob/0c2fca155941b0e411ad82a0f63f06869e070c5b/pkg/sql/parser/sql.y#L8710)):*
> ```Yacc
>     nullsOrder := $3.nullsOrder()
>     if opClass != "" {
>       if opClass == "gin_trgm_ops" || opClass == "gist_trgm_ops" {
> ```

you only added support for "gin_trgm_ops" above -- do you also want to add support for the gist version?
___
*[`pkg/sql/rowenc/index_encoding.go` line 896 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/79705#-N2dEa614TxriRZ8TNOc:-N2dEY8l6MU6nTk8bAwF:blxx1nc) ([raw file](https://github.com/cockroachdb/cockroach/blob/3c68a70d2624dcb6eca8a5850f3770bfe6a14af6/pkg/sql/rowenc/index_encoding.go#L896)):*
> ```Go
> // trigrams present in the input string.
> func EncodeLikeTrigramSpans(val *tree.DString) (inverted.Expression, error) {
> 	chunks := strings.Split(string(*val), "%")
> ```

nit: add a comment to explain why you split on %
___
*[`pkg/sql/rowenc/index_encoding.go` line 898 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/79705#-N2dDldA55c2kernVfq7:-N2dDldA55c2kernVfq8:b-uaqy87) ([raw file](https://github.com/cockroachdb/cockroach/blob/3c68a70d2624dcb6eca8a5850f3770bfe6a14af6/pkg/sql/rowenc/index_encoding.go#L898)):*
> ```Go
> 	chunks := strings.Split(string(*val), "%")
> 	// Each chunk will have at minimum 2 trigrams, so start at that default size.
> 	keys := make([][]byte, 0, len(chunks)*2)
> ```

should this be `len(chunks)+2`?
___
*[`pkg/sql/rowenc/index_encoding.go` line 897 at r17](https://reviewable.io/reviews/cockroachdb/cockroach/79705#-N2dQHgl-XnznVnqSLIi:-N2dQHgm1q0YycRkkpAc:b-iloihs) ([raw file](https://github.com/cockroachdb/cockroach/blob/0d81af2f9e6759aa40f05c5663aa49ebb3e849f4/pkg/sql/rowenc/index_encoding.go#L897)):*
> ```Go
> // expression must match every trigram in the input. Otherwise, it will match
> // any trigram in the input.
> func EncodeTrigramSpans(val *tree.DString, allMustMatch bool) (inverted.Expression, error) {
> ```

This could use a unit test (similar to `TestEncodeContainingArrayInvertedIndexSpans`, `TestEncodeOverlapsArrayInvertedIndexSpans` or `TestEncodeContainedArrayInvertedIndexSpans`).
___
*[`pkg/sql/sem/tree/create.go` line 141 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/79705#-N2dKPm24CTc8EcypWej:-N2dKPm24CTc8EcypWek:bxgsfw1) ([raw file](https://github.com/cockroachdb/cockroach/blob/e395dcb2f11619c852e7cfd0f8319765977166b8/pkg/sql/sem/tree/create.go#L141)):*
> ```Go
> 	//}
> 	//	return unimplementedWithIssue(sqllex, 47420)
> 	//}
> ```

Looks like dead code
___
*[`pkg/sql/logictest/testdata/logic_test/trigram_indexes` line 17 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/79705#-N2dLkU79JsbWgYPvX9J:-N2dLkU79JsbWgYPvX9K:b8eokdh) ([raw file](https://github.com/cockroachdb/cockroach/blob/e395dcb2f11619c852e7cfd0f8319765977166b8/pkg/sql/logictest/testdata/logic_test/trigram_indexes#L17)):*
> ```
> 
> statement ok
> CREATE INDEX ON a USING GIN(t gin_trgm_ops)
> ```

do you want to add a test for `USING GIST` and `gist_trgm_ops` too?


<!-- Sent from Reviewable.io -->

jordanlewis committed:
invertedidx: add test for trigram span creation

Release note: None

jordanlewis reviewed (commented):
Thanks for the review!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/79705#-:-N36gp08AOFz8cJfOWwi:b-9qdoz3)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rytaft)

___
*[`pkg/sql/opt/invertedidx/trigram.go` line 32 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/79705#-N2dC5GB4D9yvJrVqH52:-N36cLaPCUr2V75AhbgI:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3c68a70d2624dcb6eca8a5850f3770bfe6a14af6/pkg/sql/opt/invertedidx/trigram.go#L32)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

nit: add comment:
```
// extractInvertedFilterConditionFromLeaf is part of the invertedFilterPlanner
// interface.
```
</blockquote></details>

Done.
___
*[`pkg/sql/opt/invertedidx/trigram.go` line 32 at r17](https://reviewable.io/reviews/cockroachdb/cockroach/79705#-N2dOVgOAmKxCwd3xBz0:-N36cL7pDCR4iAc3aByw:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0d81af2f9e6759aa40f05c5663aa49ebb3e849f4/pkg/sql/opt/invertedidx/trigram.go#L32)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

This could use a unit test (similar to `TestTryFilterJsonOrArrayIndex`)
</blockquote></details>

Done.
___
*[`pkg/sql/parser/sql.y` line 8710 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/79705#-N2dJpDQEk5pRAbeEiQC:-N36cQGH8z8on53qALF3:bom8nzi) ([raw file](https://github.com/cockroachdb/cockroach/blob/0c2fca155941b0e411ad82a0f63f06869e070c5b/pkg/sql/parser/sql.y#L8710)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

you only added support for "gin_trgm_ops" above -- do you also want to add support for the gist version?
</blockquote></details>

I'm not sure. They accelerate queries differently - do we think it's misleading to pretend both are supported? This implementation is much more akin to the gin method of indexing. I think it's also questionable that we added "gist support" for JSON for the same reason, but I'm not sure.
___
*[`pkg/sql/rowenc/index_encoding.go` line 896 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/79705#-N2dEa614TxriRZ8TNOc:-N36BDb59QzkhoHen-XN:b-7gh3kg) ([raw file](https://github.com/cockroachdb/cockroach/blob/3c68a70d2624dcb6eca8a5850f3770bfe6a14af6/pkg/sql/rowenc/index_encoding.go#L896)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

nit: add a comment to explain why you split on %
</blockquote></details>

Refactored this to split outside of the function, and added comment
___
*[`pkg/sql/rowenc/index_encoding.go` line 898 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/79705#-N2dDldA55c2kernVfq7:-N36BKzo5zugkFD7Pgk3:b8o9m99) ([raw file](https://github.com/cockroachdb/cockroach/blob/3c68a70d2624dcb6eca8a5850f3770bfe6a14af6/pkg/sql/rowenc/index_encoding.go#L898)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

should this be `len(chunks)+2`?
</blockquote></details>

No, this is correct - the idea is that we want to make the array have minimum size of (2 trigrams) * (the number of chunks).
___
*[`pkg/sql/rowenc/index_encoding.go` line 897 at r17](https://reviewable.io/reviews/cockroachdb/cockroach/79705#-N2dQHgl-XnznVnqSLIi:-N36bx8d07yNKnI0h-yj:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0d81af2f9e6759aa40f05c5663aa49ebb3e849f4/pkg/sql/rowenc/index_encoding.go#L897)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

This could use a unit test (similar to `TestEncodeContainingArrayInvertedIndexSpans`, `TestEncodeOverlapsArrayInvertedIndexSpans` or `TestEncodeContainedArrayInvertedIndexSpans`).
</blockquote></details>

Done.
___
*[`pkg/sql/sem/tree/create.go` line 141 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/79705#-N2dKPm24CTc8EcypWej:-N36c4fW4Mry0MvlT6qI:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e395dcb2f11619c852e7cfd0f8319765977166b8/pkg/sql/sem/tree/create.go#L141)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

Looks like dead code
</blockquote></details>

Done.
___
*[`pkg/sql/logictest/testdata/logic_test/trigram_indexes` line 17 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/79705#-N2dLkU79JsbWgYPvX9J:-N36ddfs-xPHVlcfiyF2:b-by7caz) ([raw file](https://github.com/cockroachdb/cockroach/blob/e395dcb2f11619c852e7cfd0f8319765977166b8/pkg/sql/logictest/testdata/logic_test/trigram_indexes#L17)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

do you want to add a test for `USING GIST` and `gist_trgm_ops` too?
</blockquote></details>

See comment above - not sure we really want to be supporting them.


<!-- Sent from Reviewable.io -->

jordanlewis commented:
RFAL. I fixed a couple of issues with span creation and added more tests. The key change was that we can't generate tight spans for trigrams except for in the unusual case of a single trigram with no left or right anchors, like `LIKE '%foo'`, which seems like a borderline enough use case that I decided to just not worry about ever generating tight spans for trigram index searches.

rytaft reviewed (commented):
> Reviewed 46 of 46 files at r18, 38 of 38 files at r19, 27 of 27 files at r20, 1 of 1 files at r21, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/79705)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis and @rytaft)

___
*[`pkg/sql/parser/sql.y` line 8710 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/79705#-N2dJpDQEk5pRAbeEiQC:-N37XY_K8J2-DcHG7yDm:b-chlrwx) ([raw file](https://github.com/cockroachdb/cockroach/blob/0c2fca155941b0e411ad82a0f63f06869e070c5b/pkg/sql/parser/sql.y#L8710)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

I'm not sure. They accelerate queries differently - do we think it's misleading to pretend both are supported? This implementation is much more akin to the gin method of indexing. I think it's also questionable that we added "gist support" for JSON for the same reason, but I'm not sure.
</blockquote></details>

I think it's probably useful for compatibility reasons and for consistency with our existing support for JSON -- as long as our docs explain that under the hood they are all the same...
___
*[`pkg/sql/opt/exec/execbuilder/testdata/trigram_index` line 238 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/79705#-N37UfNP90-IYDGGNGEp:-N37UfNP90-IYDGGNGEq:b-l60prf) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf88ed5b6a358fb7d7bb4fa1356fee9d6b2d9163/pkg/sql/opt/exec/execbuilder/testdata/trigram_index#L238)):*
> ```
> EXPLAIN SELECT * FROM a WHERE b = 'foobar'
> ----
> distribution: local
> ```

looks like you need to use --rewrite for this and the one below
___
*[`pkg/sql/rowenc/index_encoding_test.go` line 927 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/79705#-N37V5xA6Pzl7nYEwm-J:-N37V5xA6Pzl7nYEwm-K:b9q7y9y) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf88ed5b6a358fb7d7bb4fa1356fee9d6b2d9163/pkg/sql/rowenc/index_encoding_test.go#L927)):*
> they are not necessarily contained by the value.

Does this apply to this test? Or is it copy pasta?
___
*[`pkg/sql/rowenc/index_encoding_test.go` line 941 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/79705#-N37VaSX6IxVc0ru1C60:-N37VaSX6IxVc0ru1C61:bx3672) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf88ed5b6a358fb7d7bb4fa1356fee9d6b2d9163/pkg/sql/rowenc/index_encoding_test.go#L941)):*
> ```Go
> 
> 		// "Reverse order" trigrams shouldn't match.
> 		{`test`, `tse`, like, false, false, true},
> ```

dup from the previous line
___
*[`pkg/sql/rowenc/index_encoding_test.go` line 976 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/79705#-N37W-8bCqZ8nuYu7s43:-N37W-8bCqZ8nuYu7s44:b-l0d894) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf88ed5b6a358fb7d7bb4fa1356fee9d6b2d9163/pkg/sql/rowenc/index_encoding_test.go#L976)):*
> ```Go
> 			expr, err = parser.ParseExpr(fmt.Sprintf("'%s' %s '%s'", indexedValue, opstr, value))
> 			require.NoError(t, err)
> 			chunks = strings.Split(value, "%")
> ```

this seems incorrect for the = case
___
*[`pkg/sql/rowenc/index_encoding_test.go` line 996 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/79705#-N37WKzvEqg6pQ7MB89_:-N37WKzvEqg6pQ7MB89a:b-crmiky) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf88ed5b6a358fb7d7bb4fa1356fee9d6b2d9163/pkg/sql/rowenc/index_encoding_test.go#L996)):*
> ```Go
> 			t.Fatalf("unexpectedly found a tight expression")
> 		}
> 		require.Equal(t, expectUnique, spanExpr.Unique, "%s, %s: unexpected unique attribute", indexedValue, value)
> ```

Do you also want to print the value of `expectUnique`?
___
*[`pkg/sql/rowenc/index_encoding_test.go` line 1002 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/79705#-N37WWFY2wwgpdKPvwrq:-N37WWFY2wwgpdKPvwrr:bzibntc) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf88ed5b6a358fb7d7bb4fa1356fee9d6b2d9163/pkg/sql/rowenc/index_encoding_test.go#L1002)):*
> ```Go
> 		require.NoError(t, err)
> 
> 		require.Equal(t, expectContainsKeys, containsKeys, "%s, %s: expected containsKeys", indexedValue, value)
> ```

Do you also want to print the value of `expectContainsKeys`?
___
*[`pkg/sql/rowenc/index_encoding_test.go` line 1013 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/79705#-N37WsxTBGa1NkiIXrEl:-N37WsxTBGa1NkiIXrEm:bswld0q) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf88ed5b6a358fb7d7bb4fa1356fee9d6b2d9163/pkg/sql/rowenc/index_encoding_test.go#L1013)):*
> ```Go
> 		require.NoError(t, err)
> 		actual := bool(*datum.(*tree.DBool))
> 		require.Equal(t, expected, actual, "%s, %s: expected evaluation result to match", indexedValue, value)
> ```

Ditto here, want to print value for `expected`?
___
*[`pkg/sql/rowenc/index_encoding_test.go` line 1020 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/79705#-N37WlPKCM7qQ8tjRqE7:-N37WlPKCM7qQ8tjRqE8:b312lbb) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf88ed5b6a358fb7d7bb4fa1356fee9d6b2d9163/pkg/sql/rowenc/index_encoding_test.go#L1020)):*
> ```Go
> 		t.Logf("test case: %v", c)
> 		runTest(c.indexedValue, c.value, c.searchType, c.containsKeys, c.expected, c.unique)
> 	}
> ```

Is there any way to add some randomized testing for this?
___
*[`pkg/sql/rowenc/index_encoding_test.go` line 1025 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/79705#-N37X2n84rMTFYB_1VCu:-N37X2n84rMTFYB_1VCv:b3za5y5) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf88ed5b6a358fb7d7bb4fa1356fee9d6b2d9163/pkg/sql/rowenc/index_encoding_test.go#L1025)):*
> ```Go
> func TestEncodeTrigramInvertedIndexSpansError(t *testing.T) {
> 	// Make sure that any input with a chunk with fewer than 3 characters returns
> 	// an error, since we can't produce
> ```

this sentence got cut off
___
*[`pkg/sql/rowenc/index_encoding_test.go` line 1035 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/79705#-N37XCLMCAH8Gmt-A-UK:-N37XCLMCAH8Gmt-A-UL:by6l6bz) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf88ed5b6a358fb7d7bb4fa1356fee9d6b2d9163/pkg/sql/rowenc/index_encoding_test.go#L1035)):*
> ```Go
> 	}
> 	for _, tc := range tcs {
> 		_, err := EncodeTrigramSpans(tc.chunks, true)
> ```

nit: `true /* allMustMatch */`


<!-- Sent from Reviewable.io -->

jordanlewis reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/79705)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rytaft)

___
*[`pkg/sql/parser/sql.y` line 8710 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/79705#-N2dJpDQEk5pRAbeEiQC:-N37cJOQ471p6OgMElNu:b-t69bp6) ([raw file](https://github.com/cockroachdb/cockroach/blob/0c2fca155941b0e411ad82a0f63f06869e070c5b/pkg/sql/parser/sql.y#L8710)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

I think it's probably useful for compatibility reasons and for consistency with our existing support for JSON -- as long as our docs explain that under the hood they are all the same...
</blockquote></details>

Fair enough! Done.
___
*[`pkg/sql/opt/exec/execbuilder/testdata/trigram_index` line 238 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/79705#-N37UfNP90-IYDGGNGEp:-N37cXi1BgjgLOJERHmT:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf88ed5b6a358fb7d7bb4fa1356fee9d6b2d9163/pkg/sql/opt/exec/execbuilder/testdata/trigram_index#L238)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

looks like you need to use --rewrite for this and the one below
</blockquote></details>

Done.
___
*[`pkg/sql/rowenc/index_encoding_test.go` line 927 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/79705#-N37V5xA6Pzl7nYEwm-J:-N37cYgJ6QKZDs_PAmKu:b-159o1) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf88ed5b6a358fb7d7bb4fa1356fee9d6b2d9163/pkg/sql/rowenc/index_encoding_test.go#L927)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

> they are not necessarily contained by the value.

Does this apply to this test? Or is it copy pasta?
</blockquote></details>

Copy pasta, done.
___
*[`pkg/sql/rowenc/index_encoding_test.go` line 941 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/79705#-N37VaSX6IxVc0ru1C60:-N37ct2X26MQz44Di-IC:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf88ed5b6a358fb7d7bb4fa1356fee9d6b2d9163/pkg/sql/rowenc/index_encoding_test.go#L941)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

dup from the previous line
</blockquote></details>

Done.
___
*[`pkg/sql/rowenc/index_encoding_test.go` line 976 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/79705#-N37W-8bCqZ8nuYu7s43:-N37d6yT4ChhXG5Dam7H:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf88ed5b6a358fb7d7bb4fa1356fee9d6b2d9163/pkg/sql/rowenc/index_encoding_test.go#L976)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

this seems incorrect for the = case
</blockquote></details>

Done.
___
*[`pkg/sql/rowenc/index_encoding_test.go` line 996 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/79705#-N37WKzvEqg6pQ7MB89_:-N37d7wa4rYqGF3OoxPm:b-nbdi99) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf88ed5b6a358fb7d7bb4fa1356fee9d6b2d9163/pkg/sql/rowenc/index_encoding_test.go#L996)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

Do you also want to print the value of `expectUnique`?
</blockquote></details>

That's automatically printed by the require.Equal statement.
___
*[`pkg/sql/rowenc/index_encoding_test.go` line 1002 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/79705#-N37WWFY2wwgpdKPvwrq:-N37dAHY2HRjoDgg8Kl3:b7exxsp) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf88ed5b6a358fb7d7bb4fa1356fee9d6b2d9163/pkg/sql/rowenc/index_encoding_test.go#L1002)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

Do you also want to print the value of `expectContainsKeys`?
</blockquote></details>

Ditto above, require.Equal prints that part already.
___
*[`pkg/sql/rowenc/index_encoding_test.go` line 1013 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/79705#-N37WsxTBGa1NkiIXrEl:-N37dCuG7_l-_V-L0eGd:b-siolcm) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf88ed5b6a358fb7d7bb4fa1356fee9d6b2d9163/pkg/sql/rowenc/index_encoding_test.go#L1013)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

Ditto here, want to print value for `expected`?
</blockquote></details>

Ditto :)
___
*[`pkg/sql/rowenc/index_encoding_test.go` line 1020 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/79705#-N37WlPKCM7qQ8tjRqE7:-N37dF0JDwNU3bv8O4-3:bk6xr7g) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf88ed5b6a358fb7d7bb4fa1356fee9d6b2d9163/pkg/sql/rowenc/index_encoding_test.go#L1020)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

Is there any way to add some randomized testing for this?
</blockquote></details>

I added some. I had to reimplement a bunch of the actual logic for trigram search in the test, but I think it's probably valuable. It caught a bug in the test harness itself and a "bug" in the trigram generation routine that wouldn't have been visible to users, so I think it was certainly valuable - thanks for the nudge!
___
*[`pkg/sql/rowenc/index_encoding_test.go` line 1025 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/79705#-N37X2n84rMTFYB_1VCu:-N37dEqu-3ioQPU-O6-d:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf88ed5b6a358fb7d7bb4fa1356fee9d6b2d9163/pkg/sql/rowenc/index_encoding_test.go#L1025)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

this sentence got cut off
</blockquote></details>

Done.
___
*[`pkg/sql/rowenc/index_encoding_test.go` line 1035 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/79705#-N37XCLMCAH8Gmt-A-UK:-N37dHQ57t8ox6LBe--J:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf88ed5b6a358fb7d7bb4fa1356fee9d6b2d9163/pkg/sql/rowenc/index_encoding_test.go#L1035)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

nit: `true /* allMustMatch */`
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

rytaft reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> just a few nits below

> Reviewed 25 of 25 files at r22, 17 of 17 files at r23, 6 of 6 files at r24, 1 of 1 files at r25, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/79705#-:-N3BQvfeElUeo-cmL2TV:b8nd49j)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @jordanlewis)

___
*[`-- commits` line 23 at r25](https://reviewable.io/reviews/cockroachdb/cockroach/79705#-N3BQgvn5Q1BNNAk87nZ):*
nit: add `gist_trgm_ops` here too
___
*[`pkg/sql/rowenc/index_encoding_test.go` line 1015 at r22](https://reviewable.io/reviews/cockroachdb/cockroach/79705#-N3BOsmf2ZUEGXmkB3u_:-N3BOsmf2ZUEGXmkB3ua:b9uq1jg) ([raw file](https://github.com/cockroachdb/cockroach/blob/bc0511eabd1ab88aec4f80d08c5d4b0e21fbe0cc/pkg/sql/rowenc/index_encoding_test.go#L1015)):*
> ```Go
> 		for _, searchType := range []trigramSearchType{like, eq, similar} {
> 			chunks, expr := makeTrigramBinOp(t, left, right, searchType)
> 			lTrigrams := trigram.MakeTrigrams(left, false /*pad*/)
> ```

super nit: `/*pad*/` -> `/* pad */`
___
*[`pkg/sql/rowenc/index_encoding_test.go` line 1022 at r22](https://reviewable.io/reviews/cockroachdb/cockroach/79705#-N3BOyVn2zEg8P0Fj58p:-N3BOyVo4SJQ2rbUN9ox:b-7ropln) ([raw file](https://github.com/cockroachdb/cockroach/blob/bc0511eabd1ab88aec4f80d08c5d4b0e21fbe0cc/pkg/sql/rowenc/index_encoding_test.go#L1022)):*
> ```Go
> 			all := true
> 			for _, chunk := range chunks {
> 				rTrigrams := trigram.MakeTrigrams(chunk, false /*pad*/)
> ```

ditto
___
*[`pkg/sql/rowenc/index_encoding_test.go` line 1055 at r22](https://reviewable.io/reviews/cockroachdb/cockroach/79705#-N3BQ9ph-AGg_i50ViT0:-N3BQ9pi0NVkew9TWBxs:b-5qevle) ([raw file](https://github.com/cockroachdb/cockroach/blob/bc0511eabd1ab88aec4f80d08c5d4b0e21fbe0cc/pkg/sql/rowenc/index_encoding_test.go#L1055)):*
> ```Go
> 					if valid {
> 						// Unique is false if there's more than 1 trigram.
> 						unique = false
> ```

I'm confused about why you also need to set `unique=false` here. Isn't the previous case enough?
___
*[`pkg/sql/rowenc/index_encoding_test.go` line 1061 at r22](https://reviewable.io/reviews/cockroachdb/cockroach/79705#-N3BPxO990PC23ZsXofv:-N3BPxO990PC23ZsXofw:b-8thsgl) ([raw file](https://github.com/cockroachdb/cockroach/blob/bc0511eabd1ab88aec4f80d08c5d4b0e21fbe0cc/pkg/sql/rowenc/index_encoding_test.go#L1061)):*
> ```Go
> 			}
> 			if !valid {
> 				_, err := EncodeTrigramSpans(chunks, true /* allMusMatch */)
> ```

nit: allMusMatch -> allMustMatch
___
*[`pkg/sql/rowenc/index_encoding_test.go` line 1077 at r22](https://reviewable.io/reviews/cockroachdb/cockroach/79705#-N3BO1h87h_c6DQu3YLP:-N3BO1h97r_o0rW7Kubf:b9wru58) ([raw file](https://github.com/cockroachdb/cockroach/blob/bc0511eabd1ab88aec4f80d08c5d4b0e21fbe0cc/pkg/sql/rowenc/index_encoding_test.go#L1077)):*
> ```Go
> 	var err error
> 	switch searchType {
> 	case like, eq:
> ```

nit: why not just have a different case for these two, and in all three cases assign a value for `opstr`? Then you can move the `ParseExpr` logic below the switch. Doesn't make much sense to me that you combine `like` and `eq`, but have a separate case for `similar`.


<!-- Sent from Reviewable.io -->

jordanlewis reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/79705)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @ajwerner and @rytaft)

___
*[`pkg/sql/rowenc/index_encoding_test.go` line 1055 at r22](https://reviewable.io/reviews/cockroachdb/cockroach/79705#-N3BQ9ph-AGg_i50ViT0:-N3KXEi6CB_eU4ODNokQ:bd7qj28) ([raw file](https://github.com/cockroachdb/cockroach/blob/bc0511eabd1ab88aec4f80d08c5d4b0e21fbe0cc/pkg/sql/rowenc/index_encoding_test.go#L1055)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

I'm confused about why you also need to set `unique=false` here. Isn't the previous case enough?
</blockquote></details>

It's required but was obfuscated, I simplified by just counting the # of valid trigrams across the chunks instead. 
___
*[`pkg/sql/rowenc/index_encoding_test.go` line 1077 at r22](https://reviewable.io/reviews/cockroachdb/cockroach/79705#-N3BO1h87h_c6DQu3YLP:-N3KXhLe7Krkw9DplxVE:br9h2dr) ([raw file](https://github.com/cockroachdb/cockroach/blob/bc0511eabd1ab88aec4f80d08c5d4b0e21fbe0cc/pkg/sql/rowenc/index_encoding_test.go#L1077)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

nit: why not just have a different case for these two, and in all three cases assign a value for `opstr`? Then you can move the `ParseExpr` logic below the switch. Doesn't make much sense to me that you combine `like` and `eq`, but have a separate case for `similar`.
</blockquote></details>

Nice simplification, done.


<!-- Sent from Reviewable.io -->

jordanlewis commented:
TFTR!

@ajwerner @rafiss, I'd like your review on commits 2 and 3, which add opclass parsing and persistence and a version gate. Thanks!

ajwerner reviewed (commented):
> Reviewed 6 of 13 files at r14, 2 of 46 files at r18, 2 of 25 files at r22, 23 of 23 files at r26, 17 of 17 files at r27, 4 of 6 files at r28.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/79705)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @ajwerner, @jordanlewis, @rafiss, and @rytaft)

___
*[`pkg/sql/create_index.go` line 351 at r27](https://reviewable.io/reviews/cockroachdb/cockroach/79705#-N3LakO-7o8jMcYXckr6:-N3LakO-7o8jMcYXckr7:bwpag58) ([raw file](https://github.com/cockroachdb/cockroach/blob/68691c258ca406dca74dd02f1c0c229a8f600785/pkg/sql/create_index.go#L351)):*
> ```Go
> 		}
> 		indexDesc.InvertedColumnKinds[0] = descpb.IndexDescriptor_TRIGRAM
> 	}
> ```

is this going to just accept an opclass for a column type which is, say, `INT8`?

For some defense in depth, can you return an assertion failure if this gets called with a type that is not inverted indexable? 

More generally, I don't see much if any protection against opclasses showing up in the grammar and then being summarily ignored. Maybe I missed it. Can you add some logic tests which make sure things like `CREATE INDEX idx ON (int_col gin_trgm_ops)` fail. F



_Code quote:_
```Go
	}
	return nil
```
___
*[`pkg/sql/catalog/descpb/structured.proto` line 349 at r27](https://reviewable.io/reviews/cockroachdb/cockroach/79705#-N3Lb9Y75Sbb0F8fciLX:-N3Lb9Y75Sbb0F8fciLY:b-u1kl2c) ([raw file](https://github.com/cockroachdb/cockroach/blob/68691c258ca406dca74dd02f1c0c229a8f600785/pkg/sql/catalog/descpb/structured.proto#L349)):*
> ```protobuf
>   // wouldn't be possible to distinguish a text inverted index that uses
>   // trigrams, vs a text inverted index that uses stemming.
>   enum InvertedIndexColumnKind {
> ```

can you define this in `catpb`? We're going to need it in the declarative schema changer which sits below `descpb`.



_Code quote:_
```protobuf
  enum InvertedIndexColumnKind {
    // DEFAULT is the default kind of inverted index column. JSON, Array, and
    // geo inverted indexes all are DEFAULT, though prior to 22.2 they had no
    // kind at all.
    DEFAULT = 0;
    // TRIGRAM is the trigram kind of inverted index column. It's only valid on
    // text columns.
    TRIGRAM = 1;
  }
```
___
*[`pkg/sql/rowenc/index_encoding_test.go` line 963 at r26](https://reviewable.io/reviews/cockroachdb/cockroach/79705#-N3L_eO66-E13USYJwCE:-N3L_eO66-E13USYJwCF:bgsoja9) ([raw file](https://github.com/cockroachdb/cockroach/blob/2c2de6485bcbd9c40861f79ca66b5b7f4eaac75c/pkg/sql/rowenc/index_encoding_test.go#L963)):*
> ```Go
> 	runTest := func(indexedValue, value string, searchType trigramSearchType,
> 		expectContainsKeys, expected, expectUnique bool) {
> 		t.Logf("test case: %s %s %v %t %t %t", indexedValue, value, searchType, expectContainsKeys, expected, expectUnique)
> ```

nit: is this better than a subtest?
___
*[`pkg/sql/rowenc/index_encoding_test.go` line 1098 at r26](https://reviewable.io/reviews/cockroachdb/cockroach/79705#-N3L_wHv4ZtIDaKJOFGX:-N3L_wHv4ZtIDaKJOFGY:b8209ob) ([raw file](https://github.com/cockroachdb/cockroach/blob/2c2de6485bcbd9c40861f79ca66b5b7f4eaac75c/pkg/sql/rowenc/index_encoding_test.go#L1098)):*
> ```Go
> 		{[]string{}},
> 	}
> 	for _, tc := range tcs {
> ```

nit: subtests


<!-- Sent from Reviewable.io -->

rytaft reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 23 of 23 files at r26, 17 of 17 files at r27, 6 of 6 files at r28, 1 of 1 files at r29, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/79705#-:-N3R9RokAKrlep_ZUR7P:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @jordanlewis and @rafiss)



<!-- Sent from Reviewable.io -->

mgartner reviewed (changes requested):
> Reviewed 6 of 13 files at r14, 2 of 46 files at r18, 2 of 25 files at r22, 23 of 23 files at r26, 17 of 17 files at r27, 6 of 6 files at r28, 1 of 1 files at r29, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/79705)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @jordanlewis and @rafiss)

___
*[`pkg/sql/opt/invertedidx/trigram.go` line 97 at r26](https://reviewable.io/reviews/cockroachdb/cockroach/79705#-N3UxAIe1UvcS9Lbn-Uu:-N3UxAIe1UvcS9Lbn-Uv:b-qich20) ([raw file](https://github.com/cockroachdb/cockroach/blob/2c2de6485bcbd9c40861f79ca66b5b7f4eaac75c/pkg/sql/opt/invertedidx/trigram.go#L97)):*
> ```Go
> 		// For a LIKE or ILIKE expr, we split the string by the % separator
> 		// character which is used in LIKE expressions to separate matches.
> 		chunks = strings.Split(s, "%")
> ```

This seems too naïve to handle escaped `%`s:

```
marcus=# select 'hello' LIKE '%ell%';
 ?column?
----------
 t
(1 row)

marcus=# select 'hello' LIKE '\%ell%';
 ?column?
----------
 f
(1 row)
```

We've run had previous bugs with escaped strings in LIKE expressions: https://github.com/cockroachdb/cockroach/pull/68289


<!-- Sent from Reviewable.io -->

jordanlewis reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/79705)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @ajwerner, @mgartner, @rafiss, and @rytaft)

___
*[`pkg/sql/create_index.go` line 351 at r27](https://reviewable.io/reviews/cockroachdb/cockroach/79705#-N3LakO-7o8jMcYXckr6:-N3f4Ary1QrZTqeMV1wf:byv70ho) ([raw file](https://github.com/cockroachdb/cockroach/blob/68691c258ca406dca74dd02f1c0c229a8f600785/pkg/sql/create_index.go#L351)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

is this going to just accept an opclass for a column type which is, say, `INT8`?

For some defense in depth, can you return an assertion failure if this gets called with a type that is not inverted indexable? 

More generally, I don't see much if any protection against opclasses showing up in the grammar and then being summarily ignored. Maybe I missed it. Can you add some logic tests which make sure things like `CREATE INDEX idx ON (int_col gin_trgm_ops)` fail. F
</blockquote></details>

Thanks for the catch. That's a good point, I was missing this. Done.
___
*[`pkg/sql/opt/invertedidx/trigram.go` line 97 at r26](https://reviewable.io/reviews/cockroachdb/cockroach/79705#-N3UxAIe1UvcS9Lbn-Uu:-N3fDXeY4kApzlfF8Uv3:bg8o8c) ([raw file](https://github.com/cockroachdb/cockroach/blob/2c2de6485bcbd9c40861f79ca66b5b7f4eaac75c/pkg/sql/opt/invertedidx/trigram.go#L97)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

This seems too naïve to handle escaped `%`s:

```
marcus=# select 'hello' LIKE '%ell%';
 ?column?
----------
 t
(1 row)

marcus=# select 'hello' LIKE '\%ell%';
 ?column?
----------
 f
(1 row)
```

We've run had previous bugs with escaped strings in LIKE expressions: https://github.com/cockroachdb/cockroach/pull/68289
</blockquote></details>

It's actually alright, but for a silly reason that I didn't think of either - the trigrams of a string do not include non-alpha characters. So the output of this method will be the same regardless of whether we split the string at all or not.

This makes all of this `chunk` stuff totally unnecessary, so I've removed it, which simplifies things significantly. Thanks for asking about this!

The other thing is that we always need to re-check the results of trigram searches when searching with more than 1 trigram because the index doesn't contain information about whether two trigrams are adjacent. So that would "save" us here as well, though I suppose not in the case where we include literal % characters in trigrams that we store or search by, but we don't do that anyway.
___
*[`pkg/sql/rowenc/index_encoding_test.go` line 963 at r26](https://reviewable.io/reviews/cockroachdb/cockroach/79705#-N3L_eO66-E13USYJwCE:-N3fHJNNExLrznnT0CB7:b-fnbjcb) ([raw file](https://github.com/cockroachdb/cockroach/blob/2c2de6485bcbd9c40861f79ca66b5b7f4eaac75c/pkg/sql/rowenc/index_encoding_test.go#L963)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

nit: is this better than a subtest?
</blockquote></details>

I was just following the convention in this file, so I'll keep it as is.
___
*[`pkg/sql/rowenc/index_encoding_test.go` line 1098 at r26](https://reviewable.io/reviews/cockroachdb/cockroach/79705#-N3L_wHv4ZtIDaKJOFGX:-N3fHSTWEW0AEXP48Qpx:b9ib5lm) ([raw file](https://github.com/cockroachdb/cockroach/blob/2c2de6485bcbd9c40861f79ca66b5b7f4eaac75c/pkg/sql/rowenc/index_encoding_test.go#L1098)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

nit: subtests
</blockquote></details>

Simplified this, and there are so few that I don't think it matters too much.


<!-- Sent from Reviewable.io -->

mgartner reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 25 of 25 files at r30, 16 of 19 files at r31, 9 of 9 files at r34, 6 of 6 files at r35, 1 of 1 files at r36, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/79705#-:-N3uLzkIBW2IzTDfO32x:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @ajwerner, @jordanlewis, and @rafiss)

___
*[`pkg/sql/opt/invertedidx/trigram.go` line 97 at r26](https://reviewable.io/reviews/cockroachdb/cockroach/79705#-N3UxAIe1UvcS9Lbn-Uu:-N3uIyso0-r5yGE_M3bZ:b-vno5pj) ([raw file](https://github.com/cockroachdb/cockroach/blob/2c2de6485bcbd9c40861f79ca66b5b7f4eaac75c/pkg/sql/opt/invertedidx/trigram.go#L97)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

It's actually alright, but for a silly reason that I didn't think of either - the trigrams of a string do not include non-alpha characters. So the output of this method will be the same regardless of whether we split the string at all or not.

This makes all of this `chunk` stuff totally unnecessary, so I've removed it, which simplifies things significantly. Thanks for asking about this!

The other thing is that we always need to re-check the results of trigram searches when searching with more than 1 trigram because the index doesn't contain information about whether two trigrams are adjacent. So that would "save" us here as well, though I suppose not in the case where we include literal % characters in trigrams that we store or search by, but we don't do that anyway.
</blockquote></details>

Interesting. Makes sense! :+1: 
___
*[`pkg/sql/rowenc/index_encoding_test.go` line 1085 at r30](https://reviewable.io/reviews/cockroachdb/cockroach/79705#-N3uJktn5xkGMM7c_Yqj:-N3uJktn5xkGMM7c_Yqk:b-wmqe7e) ([raw file](https://github.com/cockroachdb/cockroach/blob/e54f3dce7199ce42194c0280c9e80432f9a50911/pkg/sql/rowenc/index_encoding_test.go#L1085)):*
> ```Go
> 		"fo",
> 		"a",
> 		"",
> ```

What if there are 3 characters but not all are alpha characters, like `fo%`? Do we expect that to succeed or err?


<!-- Sent from Reviewable.io -->

ajwerner reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 1 of 6 files at r35.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/79705#-:-N3uySGZ4WfYaDkWvRuR:bnfp4nl)* status: :shipit: complete! 2 of 0 LGTMs obtained (and 1 stale) (waiting on @jordanlewis and @rafiss)



<!-- Sent from Reviewable.io -->

jordanlewis reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/79705)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 3 stale) (waiting on @ajwerner, @mgartner, and @rafiss)

___
*[`pkg/sql/rowenc/index_encoding_test.go` line 1085 at r30](https://reviewable.io/reviews/cockroachdb/cockroach/79705#-N3uJktn5xkGMM7c_Yqj:-N3w-RS71coq3Ml4j2Wl:bv8sle0) ([raw file](https://github.com/cockroachdb/cockroach/blob/e54f3dce7199ce42194c0280c9e80432f9a50911/pkg/sql/rowenc/index_encoding_test.go#L1085)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

What if there are 3 characters but not all are alpha characters, like `fo%'`? Do we expect that to succeed or err?
</blockquote></details>

Error. Added a test.


<!-- Sent from Reviewable.io -->

jordanlewis commented:
Thanks all for the reviews!

bors r+
